qiskit/circuit/__init__.py

```diff
@@ -54,81 +78,825 @@
+attributes.  This can lead to subtle data-coherence problems: the :class:`Qubit` and :class:`Clbit`
+objects used inside the subcircuit blocks of the control-flow ops will not necessarily be identical
+to the corresponding objects in the :class:`CircuitInstruction`.  Any code that consumes
+control-flow operations in Qiskit needs to be aware of this; within a subcircuit, you should treat
+``subcircuit.qubits[i]`` as if it were really ``outer_instruction.qubits[i]``, and so on.  You can
+generate an easy lookup table for this by doing::
+
+    cf_instruction: CircuitInstruction = ...
+    cf_operation: ControlFlowOp = cf_instruction.operation
+    for block in blocks:
+        # Mappings of "inner" qubits/clbits to the outer ones.
+        qubit_map = dict(zip(block.qubits, cf_instruction.qubits))
+        clbit_map = dict(zip(block.clbits, cf_instruction.clbits))
+
+        # ... do something with `block` ...
+
+Remember that you will need to propagate this information if you recurse into subblocks of
+control-flow operations.
+
+..
+    TODO: insert cross-ref to control-flow builder guide into below paragraph once written.
+
+All the subcircuit blocks in a :class:`ControlFlowOp` are required to contain the same numbers of
+:class:`Qubit`\ s and :class:`Clbit`\ s, referring to the same outer bits in the same order, such
+that the :class:`zip` loop given in the code block above works.  The inner-circuit :class:`Bit`
+objects do not need to be literally the same objects.  When using the control-flow builder interface
+(which, it cannot be stressed enough, is *highly* recommended for users), the builders will arrange
+that the inner bit objects *are* identical to the outer bit objects; the ``qubit_map`` in the code
+block above will always be a mapping ``{x: x}``, but if you are consuming the blocks, you should be
+prepared for the case that the mapping is required.
+
+Any :class:`ClassicalRegister`\ s used in a control-flow subcircuit must also be present in all
+containing blocks (*i.e.* any containing control-flow operations, and the outermost circuit), and
+all blocks in the same :class:`ControlFlowOp` need to contain the same registers.  Again, the
+builder interface will arrange for this to be the case (or produce an eager error if they cannot).
+
+When the low-level construction is being used the inner :class:`QuantumCircuit` blocks must
+manually close over any outer-scope :ref:`near-time classical computation variables
+<circuit-repr-near-time-classical>` that they use.  This is marked by these being in the
+:meth:`~QuantumCircuit.iter_captured_vars` iterator for that block.  Libraries constructing these
+blocks manually will need to track these captures when building control-flow circuit blocks and add
+them to the block using :meth:`~QuantumCircuit.add_capture` (or the ``captures`` constructor
+argument), but user code will typically use the control-flow builder interface, which handles this
+automatically.
```

Comment:
This is quite a lot of detail but I guess this is also the right location to put it into... :-)

Reply:
tbh I consider this amount of detail being necessary a strong indication that we didn't do a great job with the data model.

Reply:
Do you think this should be part of the documentation on the quantum circuit?

Reply:
I put it here because this `qiskit.circuit` module is where the control-flow operations are documented and the module-level documentation is the place for all the stuff about how to manually fit together different parts of the `cirucit` API.  I will mention this in the discussion on the `QuantumCircuit` page as well for sure (mostly in the discussion of the `Var`-related methods).

Reply:
Sounds good.